diffrax/_integrate.py

```diff
@@ -487,14 +487,25 @@ def maybe_inplace(i, u, x):
             return eqxi.buffer_at_set(x, i, u, pred=keep_step)
 
         def save_steps(subsaveat: SubSaveAt, save_state: SaveState) -> SaveState:
-            if subsaveat.steps:
-                ts = maybe_inplace(save_state.save_index, tprev, save_state.ts)
-                ys = jtu.tree_map(
-                    ft.partial(maybe_inplace, save_state.save_index),
-                    subsaveat.fn(tprev, y, args),
-                    save_state.ys,
+            steps = int(subsaveat.steps)
+            if steps:
+                save_step = (state.num_accepted_steps) % subsaveat.steps == 0
+                should_save = keep_step & save_step
+
+                def save_branch():
+                    fn_result = subsaveat.fn(tprev, y, args)
+                    new_ts = maybe_inplace(save_state.save_index, tprev, save_state.ts)
+                    new_ys = jtu.tree_map(
+                        lambda yi, ysi: maybe_inplace(save_state.save_index, yi, ysi),
+                        fn_result,
+                        save_state.ys,
+                    )
+                    return new_ts, new_ys
+
+                ts, ys = lax.cond(
+                    should_save, save_branch, lambda: (save_state.ts, save_state.ys)
                 )
```

Comment:
Aha, you've bumped into an XLA compiler footgun! I'm guessing it looked like the logic I described in https://github.com/patrick-kidger/diffrax/issues/625#issuecomment-2846752579 could be simplified into just a single cond.

Unfortunately this now means that we have our large buffers `new_ts` and `new_ys` passing through control flow, and when combined with autodiff this means in some cases that XLA will spuriously copy the buffers on each branch, rather than doing a conditional write. (This is pretty obscure inside baseball, I totally appreciate.)

It's for this reason that when writing to these buffers, it's important to encode the conditional-ness of the write in the `pred=...` argument used inside `maybe_inplace`.